coroutines-extensions/src/commonMain/kotlin/com/mirego/trikot/streams/flow/ReactiveFlow.kt

```diff
@@ -0,0 +1,223 @@
+package com.mirego.trikot.streams.flow
+
+import com.mirego.trikot.foundation.concurrent.AtomicReference
+import kotlinx.coroutines.AbstractCoroutine
+import kotlinx.coroutines.CancellableContinuation
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.internal.ChannelFlow
+import kotlinx.coroutines.flow.internal.SendingCollector
+import kotlinx.coroutines.handleCoroutineException
+import kotlinx.coroutines.intrinsics.startCoroutineCancellable
+import kotlinx.coroutines.suspendCancellableCoroutine
+import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+import org.reactivestreams.Subscription
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.EmptyCoroutineContext
+
+/**
+ * Transforms the given flow to a reactive specification compliant [Publisher].
+ *
+ * This function is integrated with `ReactorContext` from `kotlinx-coroutines-reactor` module,
+ * see its documentation for additional details.
+ */
+fun <T : Any> Flow<T>.asFlowPublisher(): Publisher<T> = FlowAsPublisher(this)
+
+/**
+ * Transforms the given reactive [Publisher] into [Flow].
+ * Use [buffer] operator on the resulting flow to specify the size of the backpressure.
+ * More precisely, it specifies the value of the subscription's [request][Subscription.request].
+ * `1` is used by default.
+ *
+ * If any of the resulting flow transformations fails, subscription is immediately cancelled and all in-flights elements
+ * are discarded.
+ *
+ * This function is integrated with `ReactorContext` from `kotlinx-coroutines-reactor` module,
+ * see its documentation for additional details.
+ */
+public fun <T : Any> Publisher<T>.asFlow(): Flow<T> = PublisherAsFlow(this, 1)
+
+
+/**
+ * Adapter that transforms [Flow] into TCK-complaint [Publisher].
+ * [cancel] invocation cancels the original flow.
+ */
+private class FlowAsPublisher<T : Any>(private val flow: Flow<T>) : Publisher<T> {
+    override fun subscribe(s: Subscriber<in T>) {
+        val subscription = FlowSubscription(flow, s)
+        s.onSubscribe(subscription)
+        subscription.request(Long.MAX_VALUE)
+    }
+}
+
+@UseExperimental(InternalCoroutinesApi::class)
+class FlowSubscription<T>(
+    val flow: Flow<T>,
+    val subscriber: Subscriber<in T>
+) : Subscription, AbstractCoroutine<Unit>(Dispatchers.Unconfined, false) {
```

Comment:
Using `Dispatchers.Unconfined` does not work properly on iOS for the moment, we had to change it to `CustomDispatchers.Main` in the meantime this gets fixed.

Reply:
Would be nice to include CustomDispatchers in this module also.